File: Programming Language/C#/is-graph-bipartite.cs
public class Solution {
    bool ok;
    public bool IsBipartite(int[][] graph) {
        int n = graph.Length;
        int[] color = new int[n];
        for(int i = 0; i < n; i++){
            color[i] = -1;
        }
        for(int i = 0; i < n; i++){
            if(color[i] == -1){
                ok = true;
                dfs(i, 0, color, graph);
                if(ok == false) return false;
            }
        }
        return true;
    }
    public void dfs(int u, int c, int[] color, int[][] graph){
        color[u] = c;
        foreach(int v in graph[u]){
            if(color[v] == -1){
                dfs(v, c^1, color, graph);
            }else{
                if(color[u] == color[v]){
                    ok = false;
                }
            }
        }
    }
}

File: Programming Language/C/is-graph-bipartite.c
bool ok;
void dfs(int u, int c, int *color, int** graph, int graphSize, int* graphColSize){
    color[u] = c;
    for(int i = 0; i < graphColSize[u]; i++){
        int v = graph[u][i];
        if(color[v] == -1){
            dfs(v, c^1, color, graph, graphSize, graphColSize);
        }else{
            if(color[u] == color[v]){
                ok = false;
            }
        }
    }
}
bool isBipartite(int** graph, int graphSize, int* graphColSize){
    int *color = malloc(graphSize*sizeof(int));
    for(int i = 0; i < graphSize; i++){
        color[i] = -1;
    }
    for(int i = 0; i < graphSize; i++){
        if(color[i] == -1){
            ok = true;
            dfs(i, 0, color, graph, graphSize, graphColSize);
            if(ok == false){
                return false;
            }
        }
    }
    return true;
}
